File: src/lib/api/devices.ts
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { createQuery } from "@tanstack/svelte-query";

import type {
  DeviceId,
  DeviceModel,
  DeviceRequest,
  ConnectedDevice,
  DeviceRequestId,
} from "./types/devices";

import { queryClient } from "./client";

const devicesKeys = {
  root: ["devices"],
  requests: ["devices", "requests"],
  devices: ["devices", "devices"],
  connectedDevices: ["devices", "connected"],
};

// [REQUESTS] ------------------------------------------------------

function getDeviceRequests() {
  return invoke<DeviceRequest[]>("devices_get_requests");
}

function getDevices() {
  return invoke<DeviceModel[]>("devices_get_devices");
}

function getConnectedDevices() {
  return invoke<ConnectedDevice[]>("devices_get_connected_devices");
}

export function approveDeviceRequest(requestId: DeviceRequestId) {
  return invoke<void>("devices_approve_request", { requestId });
}

export function declineDeviceRequest(requestId: DeviceRequestId) {
  return invoke<void>("devices_decline_request", { requestId });
}

export function revokeDevice(deviceId: DeviceId) {
  return invoke<void>("devices_revoke_device", { deviceId });
}

// [QUERIES] ------------------------------------------------------

export function deviceRequestsQuery() {
  return createQuery({
    queryKey: devicesKeys.requests,
    queryFn: getDeviceRequests,
  });
}

export function devicesQuery() {
  return createQuery({
    queryKey: devicesKeys.devices,
    queryFn: getDevices,
  });
}

export function connectedDevicesQuery() {
  return createQuery({
    queryKey: devicesKeys.connectedDevices,
    queryFn: getConnectedDevices,
  });
}

// [MUTATORS] ------------------------------------------------------

export function invalidateDeviceRequests() {
  queryClient.invalidateQueries({
    queryKey: devicesKeys.requests,
    exact: false,
  });
}

export function invalidateDevices() {
  queryClient.invalidateQueries({
    queryKey: devicesKeys.devices,
    exact: false,
  });
}

export function invalidateConnectedDevices() {
  queryClient.invalidateQueries({
    queryKey: devicesKeys.connectedDevices,
    exact: false,
  });
}

export function invalidateDevicesAll() {
  queryClient.invalidateQueries({
    queryKey: devicesKeys.root,
    exact: false,
  });
}

// [LISTENERS] ------------------------------------------------------

listen<DeviceRequestId>("device_requests:added", () => {
  invalidateDeviceRequests();
});

listen<DeviceRequestId>("device_requests:removed", () => {
  invalidateDeviceRequests();
});

listen<DeviceRequestId>("device_requests:accepted", () => {
  invalidateDevicesAll();
});

listen<DeviceRequestId>("device_requests:declined", () => {
  invalidateDeviceRequests();
});

listen<DeviceId>("device:authenticated", () => {
  invalidateConnectedDevices();
});

listen<DeviceId>("device:disconnected", () => {
  invalidateConnectedDevices();
});

listen<DeviceId>("device:revoked", () => {
  invalidateDevices();
  invalidateConnectedDevices();
});

File: src-tauri/src/commands/devices.rs
use std::sync::Arc;

use tauri::State;

use crate::{
    database::{
        DbPool,
        entity::{
            device::{DeviceId, DeviceModel},
            folder::FolderId,
            profile::ProfileId,
        },
    },
    device::{ConnectedDevice, DeviceRequest, DeviceRequestId, Devices},
};

use super::CmdResult;

/// Gets a list of current device approval requests
#[tauri::command]
pub fn devices_get_requests(devices: State<'_, Arc<Devices>>) -> Vec<DeviceRequest> {
    devices.get_device_requests()
}

/// Deny a specific device request
#[tauri::command]
pub async fn devices_get_devices(db: State<'_, DbPool>) -> CmdResult<Vec<DeviceModel>> {
    let devices = DeviceModel::all(db.inner()).await?;
    Ok(devices)
}

/// Deny a specific device request
#[tauri::command]
pub fn devices_get_connected_devices(devices: State<'_, Arc<Devices>>) -> Vec<ConnectedDevice> {
    devices.get_connected_devices()
}

/// Approve a specific device request
#[tauri::command]
pub async fn devices_approve_request(
    request_id: DeviceRequestId,
    devices: State<'_, Arc<Devices>>,
) -> CmdResult<()> {
    devices.approve_device_request(request_id).await?;
    Ok(())
}

/// Deny a specific device request
#[tauri::command]
pub async fn devices_decline_request(
    request_id: DeviceRequestId,
    devices: State<'_, Arc<Devices>>,
) -> CmdResult<()> {
    devices.decline_device_request(request_id)?;
    Ok(())
}

/// Deny a specific device request
#[tauri::command]
pub async fn devices_revoke_device(
    device_id: DeviceId,
    devices: State<'_, Arc<Devices>>,
) -> CmdResult<()> {
    devices.revoke_device(device_id).await?;
    Ok(())
}

/// Set the current profile for a device
#[tauri::command]
pub async fn devices_set_device_profile(
    device_id: DeviceId,
    profile_id: ProfileId,
    devices: State<'_, Arc<Devices>>,
) -> CmdResult<()> {
    devices.update_device_profile(device_id, profile_id).await?;
    Ok(())
}

/// Set the current folder for a device
#[tauri::command]
pub async fn devices_set_device_folder(
    device_id: DeviceId,
    folder_id: FolderId,
    devices: State<'_, Arc<Devices>>,
) -> CmdResult<()> {
    devices.update_device_folder(device_id, folder_id).await?;
    Ok(())
}
